runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/io/GzipSdkSourceTest.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.hashing.crc32
+import kotlinx.coroutines.test.runTest
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+
+class GzipSdkSourceTest {
+    @Test
+    fun testReadToByteArray() = runTest {
+        val payload = "Hello World"
+        val bytes = payload.encodeToByteArray()
+        val bytesHash = bytes.crc32()
+
+        val gzipSdkSource = GzipSdkSource(bytes.source())
+
+        val compressedBytes = gzipSdkSource.readToByteArray()
+        gzipSdkSource.close()
+
+        val decompressedBytes = decompressGzipBytes(compressedBytes)
+
+        assertContentEquals(bytes, decompressedBytes)
+        assertEquals(bytesHash, decompressedBytes.crc32())
+    }
+
+    @Test
+    fun testRead() = runTest {
+        val payload = "Hello World"
+        val bytes = payload.encodeToByteArray()
+        val bytesHash = bytes.crc32()
+
+        val gzipSdkSource = GzipSdkSource(bytes.source())
+        val tempBuffer = SdkBuffer()
+        val limit = 1L
+
+        while (gzipSdkSource.read(tempBuffer, limit) != -1L);
+        gzipSdkSource.close()
+
+        val compressedBytes = tempBuffer.readByteArray()
+        val decompressedBytes = decompressGzipBytes(compressedBytes)
+
+        assertContentEquals(bytes, decompressedBytes)
+        assertEquals(bytesHash, decompressedBytes.crc32())
+    }
+
+    @Test
+    fun testReadLargeBody() = runTest {
+        val payload = "Hello World".repeat(100)
+        val bytes = payload.encodeToByteArray()
+        val bytesHash = bytes.crc32()
+
+        val gzipSdkSource = GzipSdkSource(bytes.source())
+        val tempBuffer = SdkBuffer()
+        val limit = 1L
+
+        while (gzipSdkSource.read(tempBuffer, limit) != -1L);
+        gzipSdkSource.close()
+
+        val compressedBytes = tempBuffer.readByteArray()
+        val decompressedBytes = decompressGzipBytes(compressedBytes)
+
+        assertContentEquals(bytes, decompressedBytes)
+        assertEquals(bytesHash, decompressedBytes.crc32())
+    }
+
+    @Test
+    fun testReadLargeLimit() = runTest {
+        val payload = "Hello World"
+        val bytes = payload.encodeToByteArray()
+        val bytesHash = bytes.crc32()
+
+        val gzipSdkSource = GzipSdkSource(bytes.source())
+        val tempBuffer = SdkBuffer()
+        val limit = 1_000L
+
+        while (gzipSdkSource.read(tempBuffer, limit) != -1L);
+        gzipSdkSource.close()
+
+        val compressedBytes = tempBuffer.readByteArray()
+        val decompressedBytes = decompressGzipBytes(compressedBytes)
+
+        assertContentEquals(bytes, decompressedBytes)
+        assertEquals(bytesHash, decompressedBytes.crc32())
+    }
+
+    @Test
+    fun testReadLargeBodyLargeLimit() = runTest {
+        val payload = "Hello World".repeat(100)
```

Comment:
*fix*:: this is only 1100 bytes, I'd up the large body size in all these test to something greater than 16k, ideally something not evenly divisible by 1024. Internally SdkBuffer uses segment sizes of 8k IIRC so we're not even breaking into multiple segments with these sizes.